### ISSUE TITLE
compile error on Linux

### ISSUE THREAD

**Issue by tissatussa** [abulmo/amoeba#16]
i'm on Xubuntu 20.04.2 and i have this error :

$ make
ldc2 -singleobj -w -dw -mcpu=native -dip25 -preview=intpromote  -O -release -boundscheck=off -flto=full --frame-pointer=none -d-version=withPopCount -mattr=+sse4.2,+popcnt amoeba.d board.d eval.d kpk.d move.d search.d tt.d uci.d util.d weight.d
Intrinsic name not mangled correctly for type arguments! Should be: llvm.prefetch.p0i8
void (i8*, i32, i32, i32)* @llvm.prefetch
in function _D4util8prefetchFPvZv
LLVM ERROR: Broken function found, compilation aborted!
make: *** [Makefile:89: amoeba] Error 1

**Comment by abulmo** [comment]
I cannot do much about it. The ldc compiler provided with Xubuntu looks broken. I personally use the compiler provided by the ldc team: https://github.com/ldc-developers/ldc/releases/tag/v1.25.1
If you do not want to uninstall your compiler you can try to comment out the following line in the file util.d (line 82):
`	else version (LDC) llvm_prefetch(v, 0, 3, 1);`

**Comment by tissatussa** [comment]
i tried to solve the error, but i did not succeed .. i had ldc(2) installed (version 1.20.1) by apt and i uninstalled ldc by Synaptic .. then i installed ldc again by snapd and i got the (stable) version 1.24.0 -- (i don't know how to install the LDC 1.25.1 package which you pointed).

then i compiled Amoeba again, but i get this error :

$ make
ldc2 -singleobj -w -dw -mcpu=native -dip25 -preview=intpromote  -O -release -boundscheck=off -flto=full --frame-pointer=none -d-version=withPopCount -mattr=+sse4.2,+popcnt amoeba.d board.d eval.d kpk.d move.d search.d tt.d uci.d util.d weight.d
/usr/bin/ld.gold: error: /snap/ldc2/170/lib/LLVMgold-ldc.so: could not load plugin library: /snap/core/current/lib/x86_64-linux-gnu/librt.so.1: undefined symbol: __clock_nanosleep, version GLIBC_PRIVATE
/usr/bin/ld.gold: error: amoeba.o:1:3: invalid character
/usr/bin/ld.gold: error: amoeba.o:1:3: syntax error, unexpected $end
/usr/bin/ld.gold: error: amoeba.o: not an object or archive
/usr/lib/gcc/x86_64-linux-gnu/10/../../../x86_64-linux-gnu/Scrt1.o:function _start: error: undefined reference to 'main'
collect2: error: ld returned 1 exit status
Error: /usr/bin/cc failed with status: 1
make: *** [Makefile:89: amoeba] Error 1

**Comment by JohanEngelen** [comment]
> $ make
> ldc2 -singleobj -w -dw -mcpu=native -dip25 -preview=intpromote -O -release -boundscheck=off -flto=full --frame-pointer=none -d-version=withPopCount -mattr=+sse4.2,+popcnt amoeba.d board.d eval.d kpk.d move.d search.d tt.d uci.d util.d weight.d
> /usr/bin/ld.gold: error: /snap/ldc2/170/lib/LLVMgold-ldc.so: could not load plugin library: /snap/core/current/lib/x86_64-linux-gnu/librt.so.1: undefined symbol: __clock_nanosleep, version GLIBC_PRIVATE
> /usr/bin/ld.gold: error: amoeba.o:1:3: invalid character

The problem is that the linker (`ld.gold`) cannot work with plugins, which is needed for `-flto=full|thin` compilation. Because it cannot use the plugin, it cannot read the object files created by LDC with `-flto`: "invalid character".
Remove `-singleobj` and `-flto=full` from the compile command line.

**Comment by tissatussa** [comment]
@JohanEngelen  perfect ! it gave me a valid binairy without any warnings or errors.  After executing the starting command 'uci' gives this output :

```
[...]/amoeba_v3.3/src$ ./amoeba 
uci
id name Amoeba 3.3.1.l64p-l
id author Richard Delorme
option name Ponder type check default false
option name Hash type spin default 64 min 1 max 65536
option name Clear Hash type button
option name Threads type spin default 1 min 1 max 256
option name Affinity type string default 0:0
option name MultiPV type spin default 1 min 1 max 256
option name UCI_AnalyseMode type check default false
uciok
```
thank you very much !  Also the author will like this info ..

**Comment by JohanEngelen** [comment]
There is some documentation on getting ld.gold _with_ plugin support here: https://llvm.org/docs/GoldPlugin.html
Also, you can link with `lld` which has built-in LLVM LTO support.

**Comment by tissatussa** [comment]
@JohanEngelen  thanks for this info .. i leave it up to the author to understand what this means for his code .. i compiled several GitHub chess engines .. most not in D, but anyhow .. sometimes i must adjust the MakeFile a little, to run on my Xubuntu 20.04, or just install some extra modules .. not all engine creators made a detection for optimal compilation and auto set parameters .. that would be very nice to everyone, also on Linux ..

**Comment by abulmo** [comment]
So, I push a new makefile, with an help and two main options
 - fast: use pgo and lto to make the fastest possible executable
 - slow: do not use pgo an lto to compile on environments that do not support them, with the knowledge that this will make amoeba slower and thus weaker.

**Comment by tissatussa** [comment]
Again, with version 3.4 i have an error when compiling your source.
So i followed the advice above:

`Remove -singleobj and -flto=full from the compile command line`

..and it solved the issue : now i have a valid v3.4 binary !
I'm on Xubuntu 20.04.

**Comment by abulmo** [comment]
> Again, with version 3.4 i have an error when compiling your source. So i followed the advice above:
> 
> `Remove -singleobj and -flto=full from the compile command line`
> 
> ..and it solved the issue : now i have a valid v3.4 binary ! I'm on Xubuntu 20.04.

You could have compiled it with _make slow_ without removing anything from the makefile. If you wanted to keep PGO, you could have tried _make fast BUILD=slow._  
Removing -singleobj and -flto=full loses about 4% in speed ; and PGO another 2%. In strenght the respective loss is probably about 4 and 2 Elo.
I do not use (X)ubuntu, but I guess you can use the right linker by installing it and setting it to compile a fully optimized Amoeba.
